src/cmd/ksh93/sh/xec.c

```diff
@@ -3339,9 +3348,16 @@ int sh_funscope_20120720(Shell_t *shp, int argn, char *argv[], int (*fun)(void *
     shp->topscope = (Shscope_t *)prevscope;
     nv_getval(sh_scoped(shp, IFSNOD));
     shp->end_fn = 0;
-    if (nsig) memcpy((char *)&shp->st.trapcom[0], savstak, nsig);
+    if (nsig) {
+        for (isig = 0; isig < nsig; ++isig) {
+            if (shp->st.trapcom[isig] && shp->st.trapcom[isig] != Empty) {
+                free(shp->st.trapcom[isig]);
+            }
+        }
+        memcpy((char *)&shp->st.trapcom[0], savsig, nsig * sizeof(char *));
```

Comment:
The convoluted addressing and typecasting isn't needed; just `memcpy(shp->st.trapcom, ...)`.